qiskit/aqua/operators/gradients/circuit_qfis/lin_comb_full.py

```diff
@@ -0,0 +1,386 @@
+from qiskit.circuit import Gate, Qubit
+from qiskit.circuit import (QuantumCircuit, QuantumRegister, ParameterVector,
+                            ParameterExpression)
+from qiskit.circuit.library import RZGate, RXGate, HGate, XGate, SdgGate, SGate, ZGate, UGate
+
+from .circuit_qfi import CircuitQFI
+from ..circuit_gradients.lin_comb import LinComb
+from ..derivative_base import DerivativeBase
+
+
+class LinCombFull(CircuitQFI):
+    r"""Compute the full Quantum Fisher Information (QFI) given a pure, parametrized quantum state
+    with an additional working qubit and the linear combination of unitaries approach.
+
+    The QFI is:
+
+        [QFI]kl= Re[〈∂kψ|∂lψ〉−〈∂kψ|ψ〉〈ψ|∂lψ〉] * 4.
+    """
+
+    def convert(self,
+                operator: CircuitStateFn,
+                params: Optional[Union[ParameterExpression, ParameterVector,
+                                       List[ParameterExpression]]] = None,
+                ) -> ListOp:
+        r"""
+        Args:
+            operator: The operator corresponding to the quantum state |ψ(ω)〉for which we compute
+                the QFI
+            params: The parameters we are computing the QFI wrt: ω
+
+        Returns:
+            ListOp[ListOp] where the operator at position k,l corresponds to QFI_kl
+
+        Raises:
+            AquaError: If one of the circuits could not be constructed.
+            TypeError: If ``operator`` is an unsupported type.
+        """
+
+        # QFI & phase fix observable
+        qfi_observable = ~StateFn(4 * Z ^ (I ^ operator.num_qubits))
+        phase_fix_observable = ~StateFn((X + 1j * Y) ^ (I ^ operator.num_qubits))
+        # see https://arxiv.org/pdf/quant-ph/0108146.pdf
+
+        if isinstance(operator, CircuitStateFn):
+            pass
+        else:
+            raise TypeError(
+                'The gradient framework is compatible with states that are given as CircuitStateFn')
+
+        if not isinstance(params, (list, np.ndarray)):
+            params = [params]
+        state_qc = copy.deepcopy(operator.primitive)
+
+        # First, the operators are computed which can compensate for a potential phase-mismatch
+        # between target and trained state, i.e.〈ψ|∂lψ〉
+        phase_fix_states = None
+        qr_work = QuantumRegister(1, 'work_qubit')
+        work_q = qr_work[0]
+        additional_qubits: Tuple[List[Qubit], List[Qubit]] = ([work_q], [])
+        # create a copy of the original state with an additional work_q register
+        for param in params:
+            param_gates = state_qc._parameter_table[param]
+            for m, param_occurence in enumerate(param_gates):
+                coeffs_i, gates_i = LinComb._gate_gradient_dict(param_occurence[0])[
+                    param_occurence[1]]
+                for k, gate_to_insert_i in enumerate(gates_i):
+                    grad_state = QuantumCircuit(*state_qc.qregs, qr_work)
+                    grad_state.data = state_qc.data
```

Comment:
we should circumvent the `data` getter/setters since they have some validation that might slow down our code. Instead we can use copy or `_data` (copy is probably safer)
```suggestion
                    grad_state = state_qc.copy()
                    grad_state.add_register(qr_work)
```